File: src/test_runner/expect/toContain.zig
pub fn toContain(
    this: *Expect,
    globalThis: *JSGlobalObject,
    callFrame: *CallFrame,
) bun.JSError!JSValue {
    defer this.postMatch(globalThis);
    const thisValue = callFrame.this();
    const arguments_ = callFrame.arguments_old(1);
    const arguments = arguments_.slice();

    if (arguments.len < 1) {
        return globalThis.throwInvalidArguments("toContain() takes 1 argument", .{});
    }

    this.incrementExpectCallCounter();

    const expected = arguments[0];
    expected.ensureStillAlive();
    const value: JSValue = try this.getValue(globalThis, thisValue, "toContain", "<green>expected<r>");

    const not = this.flags.not;
    var pass = false;

    const ExpectedEntry = struct {
        globalThis: *JSGlobalObject,
        expected: JSValue,
        pass: *bool,
    };

    if (value.jsTypeLoose().isArrayLike()) {
        var itr = try value.arrayIterator(globalThis);
        while (try itr.next()) |item| {
            if (try item.isSameValue(expected, globalThis)) {
                pass = true;
                break;
            }
        }
    } else if (value.isStringLiteral() and expected.isStringLiteral()) {
        const value_string = try value.toSlice(globalThis, default_allocator);
        defer value_string.deinit();
        const expected_string = try expected.toSlice(globalThis, default_allocator);
        defer expected_string.deinit();

        if (expected_string.len == 0) { // edge case empty string is always contained
            pass = true;
        } else if (strings.contains(value_string.slice(), expected_string.slice())) {
            pass = true;
        } else if (value_string.len == 0 and expected_string.len == 0) { // edge case two empty strings are true
            pass = true;
        }
    } else if (try value.isIterable(globalThis)) {
        var expected_entry = ExpectedEntry{
            .globalThis = globalThis,
            .expected = expected,
            .pass = &pass,
        };
        try value.forEach(globalThis, &expected_entry, struct {
            pub fn sameValueIterator(
                _: *jsc.VM,
                _: *JSGlobalObject,
                entry_: ?*anyopaque,
                item: JSValue,
            ) callconv(.c) void {
                const entry = bun.cast(*ExpectedEntry, entry_.?);
                if (item.isSameValue(entry.expected, entry.globalThis) catch return) {
                    entry.pass.* = true;
                    // TODO(perf): break out of the `forEach` when a match is found
                }
            }
        }.sameValueIterator);
    } else {
        return globalThis.throw("Received value must be an array type, or both received and expected values must be strings.", .{});
    }

    if (not) pass = !pass;
    if (pass) return .js_undefined;

    // handle failure
    var formatter = jsc.ConsoleObject.Formatter{ .globalThis = globalThis, .quote_strings = true };
    defer formatter.deinit();
    const value_fmt = value.toFmt(&formatter);
    const expected_fmt = expected.toFmt(&formatter);
    if (not) {
        const received_fmt = value.toFmt(&formatter);
        const expected_line = "Expected to not contain: <green>{f}<r>\nReceived: <red>{f}<r>\n";
        const signature = comptime getSignature("toContain", "<green>expected<r>", true);
        return this.throw(globalThis, signature, "\n\n" ++ expected_line, .{ expected_fmt, received_fmt });
    }

    const expected_line = "Expected to contain: <green>{f}<r>\n";
    const received_line = "Received: <red>{f}<r>\n";
    const signature = comptime getSignature("toContain", "<green>expected<r>", false);
    return this.throw(globalThis, signature, "\n\n" ++ expected_line ++ received_line, .{ expected_fmt, value_fmt });
}

const bun = @import("bun");
const default_allocator = bun.default_allocator;
const strings = bun.strings;

const jsc = bun.jsc;
const CallFrame = bun.jsc.CallFrame;
const JSGlobalObject = bun.jsc.JSGlobalObject;
const JSValue = bun.jsc.JSValue;

const Expect = bun.jsc.Expect.Expect;
const getSignature = Expect.getSignature;

File: src/runtime/test_runner/expect/toContain.rs
use bun_core::strings;
use bun_jsc::{CallFrame, JSGlobalObject, JSValue, JsResult};

use super::Expect;
use super::get_signature;
use super::throw;

impl Expect {
    #[bun_jsc::host_fn(method)]
    pub(crate) fn to_contain(
        &self,
        global: &JSGlobalObject,
        frame: &CallFrame,
    ) -> JsResult<JSValue> {
        let (this, value, not) =
            self.matcher_prelude(global, frame.this(), "toContain", "<green>expected<r>")?;

        let arguments = frame.arguments();

        if arguments.len() < 1 {
            return Err(global.throw_invalid_arguments(format_args!("toContain() takes 1 argument")));
        }

        let expected = arguments[0];
        expected.ensure_still_alive();
        let mut pass = false;

        // Jest's toContain uses `===` (Array.prototype.indexOf), not Object.is:
        // `[-0]` contains `0`, `[NaN]` does not contain `NaN`.
        if value.js_type_loose().is_array_like() {
            let mut itr = value.array_iterator(global)?;
            while let Some(item) = itr.next()? {
                if item.is_strict_equal(expected, global)? {
                    pass = true;
                    break;
                }
            }
        } else if value.is_string_literal() && expected.is_string_literal() {
            let value_string = value.to_utf8(global)?;
            let expected_string = expected.to_utf8(global)?;

            if expected_string.slice().is_empty() {
                // edge case empty string is always contained
                pass = true;
            } else if strings::contains(value_string.slice(), expected_string.slice()) {
                pass = true;
            } else if value_string.slice().is_empty() && expected_string.slice().is_empty() {
                // edge case two empty strings are true
                pass = true;
            }
        } else if value.is_iterable(global)? {
            value.for_each_iter(global, |global, item| {
                let Ok(same) = item.is_strict_equal(expected, global) else {
                    return;
                };
                if same {
                    pass = true;
                    // TODO(perf): break out of the `forEach` when a match is found
                }
            })?;
        } else {
            return Err(global.throw(format_args!(
                "Received value must be an array type, or both received and expected values must be strings."
            )));
        }

        if not {
            pass = !pass;
        }
        if pass {
            return Ok(JSValue::UNDEFINED);
        }

        // handle failure
        // Each `to_fmt` borrows `&mut Formatter` for the lifetime of the returned wrapper,
        // so a second Formatter (cheap struct init, no shared state) satisfies borrowck.
        let mut formatter = super::make_formatter(global);
        let mut formatter2 = super::make_formatter(global);
        if not {
            let signature = get_signature("toContain", "<green>expected<r>", true);
            return throw!(
                this,
                global,
                signature,
                concat!(
                    "\n\n",
                    "Expected to not contain: <green>{}<r>\nReceived: <red>{}<r>\n",
                ),
                expected.to_fmt(&mut formatter),
                value.to_fmt(&mut formatter2),
            );
        }

        let signature = get_signature("toContain", "<green>expected<r>", false);
        throw!(
            this,
            global,
            signature,
            concat!(
                "\n\n",
                "Expected to contain: <green>{}<r>\n",
                "Received: <red>{}<r>\n",
            ),
            expected.to_fmt(&mut formatter),
            value.to_fmt(&mut formatter2),
        )
    }
}
